inference-engine/src/cldnn_engine/ops/tensor_iterator.cpp

```diff
@@ -0,0 +1,212 @@
+// Copyright (C) 2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "cldnn_program.h"
+#include "cldnn_common_utils.h"
+#include "cldnn_engine.h"
+
+#include <cpp/ie_cnn_network.h>
+
+#include "ngraph/op/tensor_iterator.hpp"
+#include "ngraph/op/constant.hpp"
+#include "ngraph/op/util/sub_graph_base.hpp"
+
+#include "api/loop.hpp"
+#include "api/mutable_data.hpp"
+#include "api/data.hpp"
+#include "api/reorder.hpp"
+
+#include <vector>
+#include <algorithm>
+
+using TensorIterator = ngraph::op::v0::TensorIterator;
+
+namespace CLDNNPlugin {
+
+template<class DATA_TYPE>
+static DATA_TYPE CreateIntData(Program &p, const cldnn::primitive_id& id, int32_t num) {
+    auto mem = cldnn::memory::allocate(p.GetEngine(),
+        { cldnn::data_types::i32, cldnn::format::bfyx, { 1, 1, 1, 1 } });
+    auto ptr = mem.pointer<int32_t>();
+    *ptr.begin() = num;
+    return {id, mem};
+}
+
+static cldnn::mutable_data CreateAdditionalOutputData(Program &p, const std::shared_ptr<ngraph::Node>& op,
+                                            const cldnn::primitive_id& id, const cldnn::primitive_id& input,
+                                            const int32_t output_idx) {
+    const auto precision = DataTypeFromPrecision(op->get_output_element_type(output_idx));
+    const auto format = DefaultFormatForDims(op->get_output_shape(output_idx).size());
+    const auto tensor = CldnnTensorFromIEDims(op->get_output_shape(output_idx));
+    cldnn::layout output_layout = cldnn::layout(precision, format, tensor);
+    auto mem = cldnn::memory::allocate(p.GetEngine(), output_layout);
+    auto md = cldnn::mutable_data(id, {input}, mem); // cldnn::data cannot set dependency
+    return md;
+}
+
+static void UpdateBackedge(std::vector<cldnn::loop::backedge_mapping>& back_edges,
+        const cldnn::primitive_id& old_primitive_id, const cldnn::primitive_id& new_primitive_id) {
+    for (auto& back_edge : back_edges) {
+        if (back_edge.from == old_primitive_id) {
+            back_edge.from = new_primitive_id;
+        }
+    }
+}
+
+void CreateTensorIteratorOp(Program &p, const std::shared_ptr<TensorIterator> &op) {
```

Comment:
Just out of curiosity: will it handle nested loops?

Reply:
Technically I think it can be handled, thoguth did not validated such a case. 
Nested loop will be a TensorIterator having TensorIterator in its body network. Current implementation is running body network as a separated network, so I dont see any reason of disabliing it.....

Reply:
I think the possible issue here is inputs/outputs mapping for nested tensor iterators. Please consider adding tests for such case with further TI/Loop improvements